core/unix/rseq_linux.c

```diff
@@ -0,0 +1,377 @@
+                                 PAGE_SIZE) -
+            offs;
+        str_map =
+            os_map_file(fd, &str_size, offs, NULL, MEMPROT_READ, MAP_FILE_COPY_ON_WRITE);
+        if (str_map == NULL)
+            goto rseq_process_module_cleanup;
+        strtab = (char *)(str_map + sec_hdr[elf_hdr->e_shstrndx].sh_offset - offs);
+    }
+    bool found_array = false;
+    uint i;
+    ELF_SECTION_HEADER_TYPE *sec_hdr_start = sec_hdr;
+    /* The section entries on disk need load_offs.  The rseq entries in memory are
+     * relocated and only need the offset if relocations have not yet been applied.
+     */
+    ssize_t entry_offs = 0;
+    if (at_map || (DYNAMO_OPTION(early_inject) && !dr_api_entry && !dynamo_started))
+        entry_offs = load_offs;
+    for (i = 0; i < elf_hdr->e_shnum; i++) {
+#define RSEQ_PTR_ARRAY_SEC_NAME "__rseq_cs_ptr_array"
+        if (strcmp(strtab + sec_hdr->sh_name, RSEQ_PTR_ARRAY_SEC_NAME) == 0) {
+            found_array = true;
+            byte **ptrs = (byte **)(sec_hdr->sh_addr + load_offs);
+            int j;
+            for (j = 0; j < sec_hdr->sh_size / sizeof(ptrs); ++j) {
+                /* We require that the table is loaded.  If not, bail, but unlike
+                 * failing to find section headers, make this a fatal error: better
+                 * to notify the user than try to run the rseq w/o proper handling.
+                 */
+                if (ptrs < (byte **)ma->start || ptrs > (byte **)ma->end) {
+                    REPORT_FATAL_ERROR_AND_EXIT(
+                        RSEQ_BEHAVIOR_UNSUPPORTED, 3, get_application_name(),
+                        get_application_pid(),
+                        RSEQ_PTR_ARRAY_SEC_NAME " is not in a loaded segment");
+                    ASSERT_NOT_REACHED();
+                }
+                /* We assume this is a full mapping and it's safe to read the data
+                 * (a partial map shouldn't make it to module list processing).
+                 * We do perform a sanity check to handle unusual non-relocated
+                 * cases (it's possible this array is not in a loaded segment?).
+                 */
+                byte *entry = *ptrs + entry_offs;
+                if (entry < ma->start || entry > ma->end) {
+                    REPORT_FATAL_ERROR_AND_EXIT(
+                        RSEQ_BEHAVIOR_UNSUPPORTED, 3, get_application_name(),
+                        get_application_pid(),
+                        RSEQ_PTR_ARRAY_SEC_NAME "'s entries are not in a loaded segment");
+                    ASSERT_NOT_REACHED();
+                }
+                rseq_process_entry((struct rseq_cs *)entry, entry_offs);
+                ++ptrs;
+            }
+            break;
+        }
+        ++sec_hdr;
+    }
+    if (!found_array) {
+        sec_hdr = sec_hdr_start;
+        for (i = 0; i < elf_hdr->e_shnum; i++) {
+#define RSEQ_SEC_NAME "__rseq_cs"
+#define RSEQ_OLD_SEC_NAME "__rseq_table"
```

Comment:
Maybe - at least I don't understand - why the search for the "old" name is different than the "new" name above, and why we can't share the same search function?

Reply:
The __rseq_cs_ptr_array is the recommended method, so we look for it first.  When it is used, there is *also* a section named one of these two here with the actual data.  Thus, we need to go through all the sections looking for the array only -- we do not want to walk the sections and look for both at once.  The ordering has no guarantees.

Reply:
Ok.